### PR TITLE
Modernize webpack build

### PR DESCRIPTION
**Pull request by anselmbradford**
## Additions

- Adds browserlist config file for use by babel loader.

## Changes

- Updates build process to latest `webpack`.
- Merges `gulp scripts:concat` and `gulp scripts:uglify` tasks into `gulp scripts:common`.
- Makes minified/unminified JS dependent on the `NODE_ENV` value set in the user's environment.
- Uses the same output file name (`cfpb-chart-builder.min.js`) for the generated script and the demo script.
- Fix linter errors.

## Testing

-  Run `./setup.sh`
- `gulp scripts` should pass.
- `gulp scripts:common` should pass.
- `gulp scripts:demo` should pass.
- `gulp build` should pass.
- `gulp watch` should render at http://localhost:5000/ and `test/static/cfpb-chart-builder.min.js` should be minified.
- Maps should render in IE9.
- Stop server with ctrl+c.
- Running `export NODE_ENV=development` and `gulp scripts`, the file `test/static/cfpb-chart-builder.min.js` should be unminified.
- Running `export NODE_ENV=production` and `gulp scripts`, the file `test/static/cfpb-chart-builder.min.js` should be minified.
- Running `unset NODE_ENV` and `gulp scripts`, the file `test/static/cfpb-chart-builder.min.js` should be minified.

## Notes

- We may want the generated file to be `cfpb-chart-builder.js` (without the `min`), since it seems preferable to keep the same filename when swapping between minified and unminified file, otherwise logic needs to exist in template to swap between the `.min` and non-`.min` versions.

### CODE REVIEW

**Comment by cfarm** [comment]
Do we have to make any changes upstream (like to `cfgov-refresh` or a deploy process) to add the NODE_ENV variable?

**Comment by cfarm** [comment]
I'd like to see the filename be `cfpb-chart-builder.js` when in development mode and unminified, and `cfpb-chart-builder.min.js` only when it's minified.

**Comment by anselmbradford** [comment]
> Do we have to make any changes upstream (like to cfgov-refresh or a deploy process) to add the NODE_ENV variable?

Should be okay, cfgov-refresh already uses `NODE_ENV` (e.g. https://github.com/cfpb/cfgov-refresh/blob/master/frontend.sh#L16, https://github.com/cfpb/cfgov-refresh/blob/master/travis_run.sh#L7). It does need some unrelated work in cfgov-refresh though, see [GHE]/CFGOV/platform/issues/2709.

> I'd like to see the filename be cfpb-chart-builder.js when in development mode and unminified, and cfpb-chart-builder.min.js only when it's minified.

I used to be on board with this, and it bugged me that cfgov-refresh never adhered to this. However, my feelings on it have shifted over time. The problem is if we change the filename we'd need logic in the template to switch between the two names where the script is [included in the page](https://github.com/cfpb/cfpb-chart-builder/blob/master/test/index.html#L162), which means we'd essentially be delivering debug code to production if this were delivered in a production environment (unless further logic was included to strip out the swapping code in production). More importantly, there could be all kinds of processing that could happen to a JS file that is already not conveyed in the filename (babel transpiling, resolving paths, etc).

Another way to look at it is `cfpb-chart-builder.js` is the JS file, period; it could be the source, it could be the source transpiled, it could be the source minified, etc., but which format it is depends on the context within which it appears. If it's in the `src` directory, it's the source; if it's in the `dist` directory in a development environment, it may be transpiled and unminifed; if it's in the `dist` directory in a production environment it may be transpiled and minified, depending on settings for those environments.

File: config/environment.js [hunk 1]
```diff
@@ -9,6 +9,7 @@
 const envvars = {
 
   /* eslint-disable no-process-env */
+  NODE_ENV: process.env.NODE_ENV,
```
[the remark this answers is not in the document]
Comment (anselmbradford) [reply config/environment.js]: Seems like maybe we should rename `CI_ENVIRONMENT`. I can't find anything about it other than some similar named variables in GitLab's environment. It's not a predefined variable in Travis. `NODE_ENV` on the other hand is a convention that has meaning within npm.

File: config/browserslist [hunk 2]
```diff
@@ -0,0 +1,6 @@
+# Browsers that Google Analytics Tag manager custom tags support.
+# This file (browserlist) is picked up by babel-preset-env (in webpack config).
+# See https://github.com/browserslist/browserslist#config-file
+
+> 5% in US         # Sets a threshold for possible future users.
```
[the remark this answers is not in the document]
Comment (anselmbradford) [reply config/browserslist]: See second entry [here](https://github.com/browserslist/browserslist#full-list).

Also, forgot to include `last 2 versions`. Added that!

File: config/environment.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (anselmbradford) [reply config/environment.js]: Moved to https://github.com/cfpb/cfpb-chart-builder/pull/115

**Comment by cfarm** [comment]
> which means we'd essentially be delivering debug code to production if this were delivered in a production environment (unless further logic was included to strip out the swapping code in production). More importantly, there could be all kinds of processing that could happen to a JS file that is already not conveyed in the filename (babel transpiling, resolving paths, etc).

I guess I don't see this as a problem compared to the developer-friendliness of recognizing what state the compiled JS is in depending on the name. But if you feel strongly about keeping them sharing the same name, then let's use `cfpb-chart-builder.js` instead of the `min` one.

**Comment by anselmbradford** [comment]
> I guess I don't see this as a problem compared to the developer-friendliness of recognizing what state the compiled JS is in depending on the name. But if you feel strongly about keeping them sharing the same name, then let's use cfpb-chart-builder.js instead of the min one.

Updated it. Yeah I think this is direction things are going. Google analytics scripts don't include `.min` for instance (as well as our own scripts in cfgov-refresh). Only one that includes a `.min` is modernizr, which is an outdated script anyway.

Ready for review again.

**Pull request merged**